lib/grpnames.gi

```diff
@@ -20,46 +20,575 @@
+          N,        # (possible) component of G
+          B,        # complement of G in N
+          K,        # normal subgroup
+          prodK,    # product of normal subgroups
+          Z1,       # contains a (unique) component with trivial center
+          g,a,b,    # elements of G
+          avoid,    # set of elements or indices to avoid
+          C,        # Center(G)
+          Cl,       # all conjugacy classes of G
+          Clf,      # filtered list of conjugacy classes of G
+          c1,c2,c3, # conjugacy class of G
+          com,      # true if c1 and c2 commute, false otherwise
+          prod,     # product of c1 and c2
+          RedCl,    # reducible conjugacy classes of G
+          IrrCl,    # irreducible conjugacy classes of G
+          preedges, # pre-edges of the non-commuting graph
+          edges,    # final edges of the non-commuting graph
+          e,        # one edge of the non-commuting graph
+          comp,     # components of the non-commuting graph
+          DfZ1,     # non-Abelian direct factors of Z1
+          S1;       # index set corresponding to first component
+
+    # for Abelian groups return the canonical decomposition
+    if IsTrivial(G) then
+      return [G];
+    elif IsAbelian(G) then
+      Ns := List(IndependentGeneratorsOfAbelianGroup(G),
+                                                  g -> SubgroupNC(G, [g]));
+      # for infinite groups Set may be very inefficient
+      # might cause problems because for finite groups we return a set
+      if not IsFinite(G) then
+        return Ns;
+      else
+        return Set(Ns);
+      fi;
+    fi;
+
+    if not IsFinite(G) then TryNextMethod(); fi;
+
+    # look for Abelian cyclic component from the center
+    C := Center(G);
+    # (these components cannot lie in the commutator)
+    Gd := DerivedSubgroup(G);
+    avoid := [];
+    for g in C do
+      if not g in Gd and ForAll(avoid,h -> not g in RationalClass(G, h)) then
+        N := Subgroup(G, [g]);
+        B := ComplementNormalSubgroupNC(G, N);
+        # if B is a complement to N
+        if B <> fail then
+          return Union(DirectFactorsOfGroupKN(N), DirectFactorsOfGroupKN(B));
+        # if there is no complement to N then discard the generators of N
+        else
+          AddSet(avoid, g);
+        fi;
+      fi;
+    od;
+
+    # all components are non-Abelian
+    # !!! this can (and should) be made more efficient later !!!
+    # instead of conjugacy classes we should calculate by normal subgroups
+    # generated by 1 element
+    Cl := Set(ConjugacyClasses(G));
+    RedCl := Set(Filtered(Cl, x -> Size(x) = 1));
+    Clf := Difference(Cl, RedCl);
+    preedges := [];
+    for c1 in Clf do
+      for c2 in Clf do
+        com := true;
+        prod := [];
+        if c1 <> c2 then
+          for a in c1 do
+            for b in c2 do
+              if com then
+                g := a*b;
+                if g<>b*a then
+                  com := false;
```

Comment:
Of course. I never use(d) the break command, because I am not sure which loop it breaks out of.